src/circuits/Network.stanza

```diff
@@ -0,0 +1,493 @@
+defpackage jsl/circuits/Network :
+  import core
+  import collections
+  import jitx
+  import jitx/commands
+
+  import jsl/design/introspection
+  import jsl/errors
+  import jsl/ensure
+
+doc: \<DOC>
+Abstract Network Type
+
+The circuit tree is composed of `ANT` instances that form
+a sequence of operations for the tree.
+<DOC>
+public deftype ANT
+doc: \<DOC>
+Each `ANT` instance contains a recursive set of ANT elements
+
+The elements of each ANT instance are interpreted by that node.
+Some may be empty, others may have a single instance, and others
+may have 1 or more instances.
+<DOC>
+public defmulti elements (a:ANT) -> IndexedCollection<ANT>
+
+
+val DEF_ELEM_NAME = `elem
+val DEF_ELEM_PUBLIC = false
+doc: \<DOC>
+Circuit Element for a Circuit Network
+
+This type typically includes an instantiable that will be
+instantiated in the resultant network definition.
+<DOC>
+public defstruct CircuitElement <: ANT:
+  doc: \<DOC>
+  Elements of this node - This Empty by default for `CircuitElement`
+  <DOC>
+  elements:IndexedCollection<ANT> with:
+    as-method => true
+    default => to-vector<ANT>([])
+  doc: \<DOC>
+  Element Instantiable to be created in the network
+  This instantiable type needs to define either of the following ports:
+  1.  `p[1]/p[2]` like a standard unpolarized 2-pin
+  2.  `c/a` like a standard polarized 2-pin
+
+  When the component/module has more than just these ports - it is often
+  convenient to mark it as `public` so that it can be accessed inside the
+  module.
+  <DOC>
+  elem-type:Instantiable
+  doc: \<DOC>
+  Alternative Instance Name for this Element
+  By default the elements are named `elem` but this can be c
+  confusing when constructing a complex network. This allows for the names to
+  be more explicitly defined by the user.
+  <DOC>
+  name:Symbol with:
+    default => DEF_ELEM_NAME
+  doc: \<DOC>
+  Mark's this component instance as public
+
+  This means that after the network is constructed - it will be accessible
+  by the user using dot-notation.
+  <DOC>
+  is-public?:True|False with:
+    default => DEF_ELEM_PUBLIC
+with:
+  printer => true
+  keyword-constructor => true
+
+doc: \<DOC>
+Short Helper Method for Creating `CircuitElement`
+<DOC>
+public defn Elem (
+  i:Instantiable
+  --
+  name:Symbol = DEF_ELEM_NAME,
+  is-public?:True|False = DEF_ELEM_PUBLIC
+  ) -> CircuitElement :
+  CircuitElement(elem-type = i, name = name, is-public? = is-public?)
+
+
+val DEF_TAP_PORT_ID = 1
+
+doc: \<DOC>
+Tap on a port of an Circuit Element in the Network
+
+This is a method of accessing an internal node
+of the circuit element by creating a named port.
+
+Example:
+
+Given the following circuit
+
+```
+val circ = create-circuit(a + Tap(b, `divOut`))
+```
+
+It will create a module like this:
+
+```
+p[1] -> O -- a -- O -- b -- O -> p[2]
+                  |
+                  ---> divOut
+```
+
+A and B are connected in series as normal and their
+intermediate node is extracted to a port.
+
```

Comment:
To me, so YMMV, the following would be a bit more intuitive:
```
val circ = create-circuit(a + Tap(`divOut`, b))
```
notice I flipped the order of arguments in the `Tap()` function. As I read this line from left to right, I see `a` `divOut` and then `b` which is kinda the logical order of the constructed circuit. This is just a suggestion and we return you to the regularly scheduled program.

Reply:
Ya- I can see the argument. I had `ShuntTo` as well, and it made a little more sense as: 

```
val circ = create-circuit(A + (B | ShuntTo(C, `COMMON)))
```

Where `p[2]` of `C` gets connected to `COMMON` port. But we can make the `Tap` the other way. It will be more consistent with the default for `Tap` with is to make the tap port connect to `p[1]`